File: resources/recipes/authentication/two-factor-authentication/totp-service.js
/**
 * totp-service.js — TOTP enrollment + verification service (Node/Express-ready).
 *
 * Same three pieces as the Python version:
 *   enroll()  → secret + provisioning URI + QR data-url
 *   confirm() → verify first code before enabling
 *   verify()  → login-time check with attempt limiting + backup codes
 *
 * Deps: npm install otplib qrcode
 *
 * Secrets are encrypted at rest with AES-256-GCM using a key from the
 * environment (TOTP_ENCRYPTION_KEY, 32-byte hex). Never store plaintext.
 */

const { authenticator } = require('otplib');
const QRCode = require('qrcode');
const crypto = require('crypto');

const MAX_ATTEMPTS = 5;
const ATTEMPT_WINDOW_MS = 5 * 60 * 1000;

class TOTPService {
  constructor(issuer = 'MyApp') {
    this.issuer = issuer;
    const keyHex = process.env.TOTP_ENCRYPTION_KEY;
    // Demo fallback — production MUST supply a 32-byte key via env/KMS
    this.key = keyHex ? Buffer.from(keyHex, 'hex') : crypto.randomBytes(32);
    this.users = new Map(); // userId -> { encryptedSecret, pendingSecret, enabled, backupHashes, attempts }
  }

  _user(userId) {
    if (!this.users.has(userId)) {
      this.users.set(userId, { enabled: false, backupHashes: new Set(), attempts: [] });
    }
    return this.users.get(userId);
  }

  _encrypt(plaintext) {
    const iv = crypto.randomBytes(12);
    const cipher = crypto.createCipheriv('aes-256-gcm', this.key, iv);
    const enc = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()]);
    return Buffer.concat([iv, cipher.getAuthTag(), enc]).toString('base64');
  }

  _decrypt(payload) {
    const raw = Buffer.from(payload, 'base64');
    const decipher = crypto.createDecipheriv('aes-256-gcm', this.key, raw.subarray(0, 12));
    decipher.setAuthTag(raw.subarray(12, 28));
    return Buffer.concat([decipher.update(raw.subarray(28)), decipher.final()]).toString('utf8');
  }

  _hash(code) {
    return crypto.createHash('sha256').update(code).digest('hex');
  }

  // ---------- enrollment ----------

  async enroll(userId, email) {
    const secret = authenticator.generateSecret();
    const uri = authenticator.keyuri(email, this.issuer, secret);
    const qrDataUrl = await QRCode.toDataURL(uri);
    this._user(userId).pendingSecret = this._encrypt(secret);
    return { provisioningUri: uri, qrDataUrl };
  }

  confirm(userId, token) {
    const u = this._user(userId);
    if (!u.pendingSecret) throw new Error('no pending enrollment');
    const secret = this._decrypt(u.pendingSecret);
    if (!authenticator.verify({ token, secret, window: 1 })) {
      return { enabled: false };
    }
    u.encryptedSecret = u.pendingSecret;
    u.pendingSecret = null;
    u.enabled = true;
    const codes = Array.from({ length: 10 }, () =>
      crypto.randomBytes(4).toString('hex').toUpperCase()
    );
    u.backupHashes = new Set(codes.map((c) => this._hash(c)));
    return { enabled: true, backupCodes: codes }; // show ONCE
  }

  // ---------- verification ----------

  verify(userId, token) {
    const u = this._user(userId);
    if (!u.enabled) throw new Error('2FA not enabled');

    const now = Date.now();
    u.attempts = u.attempts.filter((t) => now - t < ATTEMPT_WINDOW_MS);
    if (u.attempts.length >= MAX_ATTEMPTS) return false; // endpoint returns 429
    u.attempts.push(now);

    const hashed = this._hash(token.trim().toUpperCase());
    if (u.backupHashes.has(hashed)) {
      u.backupHashes.delete(hashed); // single-use
      return true;
    }

    const secret = this._decrypt(u.encryptedSecret);
    return authenticator.verify({ token, secret, window: 1 });
  }

  disable(userId) {
    this.users.set(userId, { enabled: false, backupHashes: new Set(), attempts: [] });
  }
}

module.exports = { TOTPService };

File: resources/recipes/authentication/two-factor-authentication/totp_service.py
"""totp_service.py — TOTP enrollment + verification service (Python).

Covers the three server-side pieces from the recipe:
  - enroll()    → secret + provisioning URI + QR (data-url)
  - confirm()   → verify first code before marking 2FA enabled
  - verify()    → login-time check with attempt rate limiting
  - backup codes → generated at enrollment, stored hashed, single-use

Deps: pip install pyotp qrcode pillow

Secrets are encrypted at rest with Fernet (AES-128-CBC + HMAC; swap for
AES-256-GCM or your KMS in production). The encryption key comes from the
environment — never from the database.
"""

import base64
import hashlib
import io
import os
import secrets
import time
from collections import defaultdict
from dataclasses import dataclass, field

import pyotp
import qrcode
from cryptography.fernet import Fernet


@dataclass
class UserTOTP:
    encrypted_secret: bytes | None = None
    enabled: bool = False
    pending_secret: bytes | None = None  # set at enroll, moved to encrypted_secret at confirm
    backup_hashes: set = field(default_factory=set)
    attempts: list = field(default_factory=list)


class TOTPService:
    MAX_ATTEMPTS = 5
    ATTEMPT_WINDOW_SEC = 300  # 5 tries per 5 minutes

    def __init__(self, issuer: str = "MyApp"):
        self.issuer = issuer
        key = os.environ.get("TOTP_ENCRYPTION_KEY")
        if not key:
            # Demo only — in production this MUST come from KMS/vault/env
            key = Fernet.generate_key().decode()
        self.fernet = Fernet(key.encode())
        self.users: dict = defaultdict(UserTOTP)

    # ---------- enrollment ----------

    def enroll(self, user_id: str, email: str) -> dict:
        """Step 1: generate secret + QR. Does NOT enable 2FA yet."""
        secret = pyotp.random_base32()
        uri = pyotp.totp.TOTP(secret).provisioning_uri(name=email, issuer_name=self.issuer)
        img = qrcode.make(uri)
        buf = io.BytesIO()
        img.save(buf, format="PNG")
        qr_b64 = base64.b64encode(buf.getvalue()).decode()

        u = self.users[user_id]
        u.pending_secret = self.fernet.encrypt(secret.encode())
        return {"provisioning_uri": uri, "qr_png_b64": qr_b64}

    def confirm(self, user_id: str, token: str) -> dict:
        """Step 2: verify the first code, then enable + issue backup codes."""
        u = self.users[user_id]
        if not u.pending_secret:
            raise ValueError("no pending enrollment")
        secret = self.fernet.decrypt(u.pending_secret).decode()
        if not pyotp.TOTP(secret).verify(token, valid_window=1):
            return {"enabled": False}

        u.encrypted_secret = u.pending_secret
        u.pending_secret = None
        u.enabled = True
        codes = [secrets.token_hex(4).upper() for _ in range(10)]
        u.backup_hashes = {self._hash_code(c) for c in codes}
        return {"enabled": True, "backup_codes": codes}  # show ONCE, then discard

    # ---------- verification ----------

    def verify(self, user_id: str, token: str) -> bool:
        """Login-time check. Rate-limited; backup codes are single-use."""
        u = self.users[user_id]
        if not u.enabled:
            raise ValueError("2FA not enabled")

        now = time.time()
        u.attempts = [t for t in u.attempts if now - t < self.ATTEMPT_WINDOW_SEC]
        if len(u.attempts) >= self.MAX_ATTEMPTS:
            return False  # 429 in the endpoint layer
        u.attempts.append(now)

        # Backup codes work like a token but burn on use
        if self._hash_code(token.strip().upper()) in u.backup_hashes:
            u.backup_hashes.discard(self._hash_code(token.strip().upper()))
            return True

        secret = self.fernet.decrypt(u.encrypted_secret).decode()
        return pyotp.TOTP(secret).verify(token, valid_window=1)

    def disable(self, user_id: str) -> None:
        """Requires re-auth upstream; here we just clear the state."""
        self.users[user_id] = UserTOTP()

    @staticmethod
    def _hash_code(code: str) -> str:
        return hashlib.sha256(code.encode()).hexdigest()
